File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        if(!head){
            return NULL;
        }

        ListNode* slow = head;
        ListNode* fast = head;
        ListNode* point = head;

        while(fast->next && fast->next->next){
            slow = slow->next;
            fast = fast->next->next;;

            if(slow == fast){
                while(point!=slow){
                    point=point->next;
                    slow=slow->next;
                }
               return point;
            }

        }
        return NULL;
    }
};

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if(head==null || head.next==null) return null;

        ListNode s = head;
        ListNode f = head;

        while(f.next!=null && f.next.next!=null){
            s = s.next;
            f = f.next.next;

            if(s==f){
                ListNode d = head;
                while(s!=d){
                    s = s.next;
                    d = d.next;
                }
                return d;
            }
        }
        return null;
    }
}